### ISSUE TITLE
Jwt schema type mismatch

### ISSUE THREAD

**Issue by xorander00** [graphile/postgraphile#1144]
I'm submitting a ...

* [ ] bug report
* [ ] feature request
* [x] question

PostGraphile version:  **4.4.3**  
  
----------  
  
The generated schema has a Jwt type which maps to the corresponding Postgres composite type defined by the developer.  However, the actual return value is a base64-encoded string which is expected to be decoded by the client & typed/mapped using the schema to the actual type.  
  
The issue is that when using the schema with other tools, they screw up because the actual payload that's delivered is not exactly what the schema says it should be.  
  
In my case, I'm working on an Android app in Kotlin.  There's a GraphQL plugin for IntelliJ / Android Studio that reads a schema and generates strongly-typed interfaces & classes from it.  It ended up expecting the value that's returned from my authentication method to be the actual Jwt type defined in the schema, fields and all.  So the tooling fails at compile-time if there is no field sub-selection on the returned payload's jwt field.  Of course, when the sub-selection is added in, it throws an error at runtime because it got a string back instead of an object that it expected based on the schema.
  
I did a quick hacky fix by just modifying the schema that PostGraphile generated so that the return type from my authentication mutation is a string.  It works fine and just requires a small bit of manual decoding of the return base64-encoded string.  Not a huge deal, especially considering there are plenty of libraries that handle this pretty well, but obviously not ideal, especially since we're talking about security code here.
  
I can think of the following options for addressing this issue:  
  
1. Add a CLI flag (e.g. --jwt-schema-string) to return a string instead of the Jwt type.  
  
2. Add a CLI flag (e.g. --jwt-schema-dual) that causes the mutation to return an object with two fields: one field that holds the base64 string token & the other field which is the json token.  
  
Option 2 seems like the better way to go.  The existing Jwt type could be changed so that it has a str field for what it currently returns and a obj field to hold the JSON.  It would have to be opt-in of course so as to not break existing expectations.  
  
Am I missing something?  Thoughts?

**Comment by benjie** [comment]
You should be able to achieve the latter with makeExtendSchemaPlugin.

**Comment by benjie** [comment]
The JWT type is a string (a signed JWT), I believe we also detail this in the generated documentation for this type. My opinion is we’re doing the right thing; could you add some code samples for queries+schema to make it more obvious where you feel we’re doing the wrong thing, I’m not sure I fully understand.

**Comment by benjie** [comment]
I'm pretty sure you've got two different PostGraphile schemas here; one that uses `--jwt-token-identifier` and one that does not. This is why you've got the inconsistencies you describe.

FYI, our schemas are generated using graphql-js (the reference GraphQL implementation) so you can be pretty sure they conform to the spec :)

How are you generating the schema dump vs executing the queries?

**Comment by benjie** [comment]
[semi-automated message] Thanks for your question; hopefully we're well on the way to helping you solve your issue. This doesn't currently seem to be a bug in the library so I'm going to close the issue, but please feel free to keep requesting help below and if it does turn out to be a bug we can definitely re-open it 👍

You can also ask for help in the #help-and-support channel in [our Discord chat](http://discord.gg/graphile).

**Comment by benjie** [comment]
If it helps, here's the schema produced [when `jwtPgTypeIdentifier` is used](https://github.com/graphile/graphile-engine/blob/87e132d7e988e5b2fe210eb74b83edc7a79aeecc/packages/postgraphile-core/__tests__/integration/schema/jwt.test.js#L7):

[snapshot](https://github.com/graphile/graphile-engine/blob/87e132d7e988e5b2fe210eb74b83edc7a79aeecc/packages/postgraphile-core/__tests__/integration/schema/__snapshots__/jwt.test.js.snap#L589-L593)

```graphql
"""
A JSON Web Token defined by [RFC 7519](https://tools.ietf.org/html/rfc7519)
which securely represents claims between two parties.
"""
scalar JwtToken
```

and here's what's produced [when it isn't](https://github.com/graphile/graphile-engine/blob/87e132d7e988e5b2fe210eb74b83edc7a79aeecc/packages/postgraphile-core/__tests__/integration/schema/defaultOptions.test.js#L6):

[snapshot](https://github.com/graphile/graphile-engine/blob/87e132d7e988e5b2fe210eb74b83edc7a79aeecc/packages/postgraphile-core/__tests__/integration/schema/__snapshots__/defaultOptions.test.js.snap#L3314-L3320)

```graphql
type JwtToken {
  a: Int
  b: BigFloat
  c: BigInt
  exp: Int
  role: String
}
```

I believe this is what you're seeing since you have 

```json
  "type": {
    "kind": "OBJECT",
    "name": "Jwt",
    "ofType": null
  },
```

It looks like an object `type` rather than a `scalar`.

**Comment by xorander00** [comment]
I'll double-check here shortly.  I happen to be working on this today, so good timing.  That scalar type that you mentioned above is what I'm looking for, so let me try again using a clean environment & see if I screwed up something before.
  
I was using "--export-schema-json schema.json --export-schema-graphql schema.graphql --sort-export --no-server" to generate the output.  I was expecting the command to dump the schemas & then exit, but it didn't do that for me.  Maybe I misinterpreted the "--no-server" flag.  I had to Ctrl-C to exit, but the dumped schema was fine.

**Comment by benjie** [comment]
If you're in `--watch` mode than `--no-server` won't make it exit because it'll keep watching for changes and writing them out each time. If you're not using `--watch` then it should definitely exit.

You must ensure your CLI flags are virtually identical (if not actually identical) between the command you use to dump and the command you use to run the server.

**Comment by xorander00** [comment]
They're identical, I just checked again.
  
No watch mode when dumping the schema either.

**Comment by xorander00** [comment]
Got the scalar when using the clean environment.  Must be some kind of misconfiguration or possibly an issue with the combination of parameters, although I'm guessing more likely the former than the latter.  
  
It also exited this time when dumping the schema.
  
I'll re-create the main environment later today & see if I run into the issue again.

**Comment by benjie** [comment]
Perhaps you've got a `.postgraphilerc` or similar causing issues?

**Comment by xorander00** [comment]
I do have a .postgraphilerc.js, but the only thing in it is the connection string.  Also, I checked the version I'm using, and both the main and clean environments are using 4.4.3.
  
Plugins could cause an issue, correct?  I do have a few that I'm using.  I'm going to re-create from empty & add plugins/params in until it starts doing it again.

**Comment by benjie** [comment]
The issue you describe with the Jwt type should only happen if `--jwt-token-identifier` is not supplied.
